Framework/Geometry/src/Instrument/ComponentInfo.cpp

```diff
@@ -183,5 +197,136 @@ double ComponentInfo::solidAngle(const size_t componentIndex,
+                absoluteBB.xMax(), absoluteBB.yMax(), absoluteBB.zMax());
+
+  // Shift
+  const Eigen::Vector3d localPos = m_componentInfo->position(index);
+  absoluteBB.xMin() += localPos[0];
+  absoluteBB.xMax() += localPos[0];
+  absoluteBB.yMin() += localPos[1];
+  absoluteBB.yMax() += localPos[1];
+  absoluteBB.zMin() += localPos[2];
+  absoluteBB.zMax() += localPos[2];
+
+  if (reference && !reference->isAxisAligned()) { // copy coordinate system
+
+    std::vector<Kernel::V3D> coordSystem;
+    coordSystem.assign(reference->getCoordSystem().begin(),
+                       reference->getCoordSystem().end());
+
+    // realign to reference coordinate system
+    absoluteBB.realign(&coordSystem);
+  }
+  return absoluteBB;
+}
+
+/**
+ * Compute the bounding box for the component with componentIndex taking into
+ *account
+ * all sub components.
+ *
+ * @param componentIndex : Component index to get the bounding box for
+ * @param reference : Optional reference for coordinate system for non-axis
+ *aligned bounding boxes
+ * @return Absolute bounding box
+ */
+BoundingBox ComponentInfo::boundingBox(const size_t componentIndex,
+                                       const BoundingBox *reference) const {
+  if (isDetector(componentIndex)) {
+    return componentBoundingBox(componentIndex, reference);
+  }
+  BoundingBox absoluteBB;
+  auto rangeComp = m_componentInfo->componentRangeInSubtree(componentIndex);
+  std::stack<std::pair<size_t, size_t>> detExclusions{};
+  auto compIterator = rangeComp.rbegin();
+  while (compIterator != rangeComp.rend()) {
+    const size_t index = *compIterator;
+    if (hasSource() && index == source()) {
+      ++compIterator;
+    } else if (isRectangularBank(index)) {
+      auto innerRangeComp = m_componentInfo->componentRangeInSubtree(index);
+      // nSubComponents, subtract off self hence -1.
+      auto nSubComponents = innerRangeComp.end() - innerRangeComp.begin() - 1;
+      auto innerRangeDet = m_componentInfo->detectorRangeInSubtree(index);
+      auto nSubDetectors =
+          std::distance(innerRangeDet.begin(), innerRangeDet.end());
+      auto nY = nSubDetectors / nSubComponents;
+      size_t corner1 = *innerRangeDet.begin();
+      size_t corner2 = corner1 + nSubDetectors - 1;
+      size_t corner3 = corner1 + (nY - 1);
+      size_t corner4 = corner2 - (nY - 1);
+
+      absoluteBB.grow(componentBoundingBox(corner1, reference));
+      absoluteBB.grow(componentBoundingBox(corner2, reference));
+      absoluteBB.grow(componentBoundingBox(corner3, reference));
+      absoluteBB.grow(componentBoundingBox(corner4, reference));
+
+      // Get bounding box for rectangular bank.
+      // Record detector ranges to skip
+      // Skip all sub components.
+      detExclusions.emplace(std::make_pair(corner1, corner2));
+      compIterator = innerRangeComp.rend();
+    } else {
+      absoluteBB.grow(componentBoundingBox(index, reference));
+      ++compIterator;
+    }
```

Comment:
Same here, I suppose this is just copied from the old method to compute the BB?

Reply:
The `RectangularDetector` class used to have its own `getBoundingBox` method that filled in the details for itself. If we have other detector types this `if/else if` is going to get a bit long isn't it? Could the visitor pattern work here too?

Reply:
I'll look into it.